markdown: Add support for online mentions.

akshatdalton opened the pull request:
This commit adds support for online mentions, `@**online**/@**here**`. Users can mention all the other online users. This sends a desktop notification to the users who are subscribed to it.

Online mention policy is added so that this feature doesn't become spammy for other users.

Detailed test cases are added for the new flag: 'online_mentioned'.

Fixes #2183.

akshatdalton commented:
Online mention policy is added as `online_mentioned` flag works differently than `wildcard_mentioned` flag. This feature is more light weight than wildcard mention, in terms of -
* number of users to mention at a time
* no mobile and email push notifications
This just sends desktop notification to the online users who are subscribed to it (if wildcard_mention_notify=True).
So admins may want to configure it differently than wildcard mentions.

akshatdalton commented:
While doing this project I found following issues:
1. As soon as a user logs out of Zulip, `UserPresence` is not updated. This shows that user is online (unless ONLINE_THRESHOLD_SECS=140 is not crossed), i.e. user_green_circle doesn't go away. I think this should be updated as soon as user is logged out.
2. Silent mentions doesn't work for wildcards
3. Wildcard mention in a blockquote sends all notifications and a normal user can't write a message as: 
```
```quote/python
... something .... @**all** ...something...
```

I think above should be resolved as well to improve user experience :).

akshatdalton commented:
And also please tell me if this feature needs more test coverage :).

timabbott commented on zerver/migrations/0307_realm_online_mention_policy.py:
```diff
@@ -0,0 +1,18 @@
+            name="online_mention_policy",
+            field=models.PositiveSmallIntegerField(default=4),
+        ),
+    ]
```
The migrations file changes in this PR aren't correct; see https://zulip.readthedocs.io/en/latest/subsystems/schema-migrations.html for how to renumber your migrations properly when rebasing (which I assume is what went wrong here).

timabbott commented on zerver/migrations/0177_user_message_add_and_index_is_private_flag.py:
```diff
@@ -32,7 +32,7 @@ def reset_is_private_flag(apps: StateApps, schema_editor: DatabaseSchemaEditor)
             flag_set_objects = (
                 UserMessage.objects.filter(user_profile__id=user_id)
-                .extra(where=["flags & 2048 != 0"])
+                .extra(where=["flags & 4096 != 0"])
```
We can't readily renumber the `is_private` bit to a different slot; instead, we always add new bits at the end of the bitword.

timabbott commented on zerver/migrations/0067_archived_models.py:
```diff
@@ -105,6 +105,7 @@ class Migration(migrations.Migration):
                             "collapsed",
                             "mentioned",
                             "wildcard_mentioned",
+                            "online_mentioned",
```
Note also that it doesn't make sense to modify old migrations; the schema migrations docs should explain this, but if not, read about Django's migrations system.  The whole point is that migrations files are static once implemented so that you can apply a series of them to change the database from its original format from a previous version to the modern format.

timabbott commented:
@akshatdalton thanks for working on this!  I posted a batch of initial comments that should help you clean up this PR to be more readily reviewable for the details.

akshatdalton commented:
@timabbott I have cleaned up the PR, specifically the Migration area that you mentioned. Please review.

akshatdalton committed:
markdown: Add support for online mentions.

This commit adds support for online mentions, @**online**/@**here**.
Users can mention all the other online users. This sends a desktop
notification to the users who are subscribed to it.

Online mention policy is added so that this feature doesn't become
spammy for other users.

Fixes #2183.

akshatdalton committed:
markdown: Add Frontend tests for online mention feature.

Detailed frontend test cases are added for the new
flag: 'online_mentioned'.

akshatdalton committed:
markdown: Add Backend tests for online mention feature.

Detailed backend test cases are added for the new
flag: 'online_mentioned'.
